core/src/inputs/advice/merkle_path_set.rs

```diff
@@ -0,0 +1,305 @@
+    }
+
+    /// Replaces the leaf at the specified index with the provided value.
+    ///
+    /// # Errors
+    /// Returns an error if the node with specified index is not in the map.
+    pub fn update_leaf(&mut self, index: u64, value: Word) -> Result<(), AdviceSetError> {
+        let depth = self.total_depth;
+        if index >= 2u64.pow(depth) {
+            return Err(AdviceSetError::InvalidIndex(depth, index));
+        }
+
+        let mut pos = 2u64.pow(depth) + index;
+
+        // Fill old_hashes vector -----------------------------------------------------------------
+        let path = self.paths.get(&(pos / 2)).unwrap();
+        let mut path_iter = path.iter();
+
+        let mut old_hashes = Vec::<Word>::new();
+
+        let mut path_hash;
+        let mut old_hash;
+
+        if is_even(pos) {
+            old_hash = *path_iter.next().unwrap();
+            path_hash = *path_iter.next().unwrap();
+        } else {
+            path_hash = *path_iter.next().unwrap();
+            old_hash = *path_iter.next().unwrap();
+        }
+
+        for _ in 0..depth - 2 {
+            old_hash = calculate_parent_hash(old_hash, pos, path_hash);
+            pos /= 2;
+            old_hashes.push(old_hash);
+            path_hash = *path_iter.next().unwrap();
+        }
+
+        // Fill new_hashes vector -----------------------------------------------------------------
+        let mut pos = 2u64.pow(depth) + index;
+
+        let mut new_hashes = Vec::<Word>::new();
+
+        let mut new_hash;
+        let mut path_hash;
+
+        if is_even(pos) {
+            self.paths.get_mut(&(pos / 2)).unwrap()[0] = value;
+        } else {
+            self.paths.get_mut(&(pos / 2)).unwrap()[1] = value;
+        }
+
+        let path = self.paths.get(&(pos / 2)).unwrap();
+        let mut path_iter = path.iter();
+
+        if is_even(pos) {
+            new_hash = *path_iter.next().unwrap();
+            path_hash = *path_iter.next().unwrap();
+        } else {
+            path_hash = *path_iter.next().unwrap();
+            new_hash = *path_iter.next().unwrap();
+        }
+
+        for _ in 0..depth - 2 {
+            new_hash = calculate_parent_hash(new_hash, pos, path_hash);
+            pos /= 2;
+            new_hashes.push(new_hash);
+            path_hash = *path_iter.next().unwrap();
+        }
+
+        // update root ----------------------------------------------------------------------------
+        self.root = calculate_parent_hash(new_hash, 1, path_hash);
+
+        // update paths ---------------------------------------------------------------------------
+        for (_, val) in self.paths.iter_mut() {
+            'path: for path_hash in val.iter_mut() {
+                for j in 0..old_hashes.len() {
+                    if *path_hash == old_hashes[j] {
+                        *path_hash = new_hashes[j];
+                        break 'path;
+                    }
+                }
+            }
+        }
```

Comment:
it feels like there is one loop too many here. You need to iterate over paths, that's one loop. Then, for each path you can just iterate over its nodes in reverse order and compare them to old hashes in the same position. And if the hashes match replace them with the new node values. Iterating over `old_hashes` seems unnecessary. Or am I missing something?